cip/CIP2015-09-16-public-type-system-type-annotation.adoc

```diff
@@ -0,0 +1,409 @@
+A Cypher user commonly should not be required to be aware of internal types though they may influence behavior or become visible in error messages.
+Public types contain less information than internal types.
+They can be specified in type annotations and there should exist straightforward mappings between the public type system and type systems of target languages of officially supported drivers and the store.
+
+The diagram below gives an overview on the various concepts around the Cypher type system and how they relate to each other as seen by this CIP.
+
+image:CIP2015-09-16-public-type-system-overview.png[Public Type System Overview]
+
+=== Type Mapping
+The issue of mapping types from the public type system to other type systems (such as Java, JavaScript or PackStream) is not the concern of this document.
+While still requiring clear definition, such mapping definitions are not a Cypher language concern.
+
+=== Type Aliases
+Some of the type names are very long and thus could be somewhat difficult to type, and perhaps even read in longer function signatures.
+Introducing a set of predefined and perhaps user defined type aliases, such as REL for RELATIONSHIP might be a remedy to this issue.
+
+=== Possible Future Use of Type Syntax
+
+==== Type annotations in declarations and definitions
+It may be helpful to extend definitions that introduce new identifiers (WITH, RETURN) or signatures in procedure declarations with type annotations for improved type checking, planning, and possibly code generation.
+
+*Example*
+`RETURN expr AS a :: NUMBER`
+
+==== Type ascription ("safe upcast")
+A type ascription annotates an expression with a type such that casting the value at runtime to the given type cannot fail.
+This may be useful (and in fact is used by Scala for example) to control what types are inferred by a type inferencing algorithm during semantic checking.
+
+*Example*
+`RETURN [1.0, 2.3, 3.0] :: LIST OF NUMBER`
+
+==== Type cast ("unsafe downcast")
+Beyond type ascription, a type cast asserts a specific type for an expression such that it cannot be verified at compile time if evaluating that expression will always produce a value of the target type.
+Hence type casts may fail at runtime.
+
+*Example*
+`RETURN CAST n.prop :: NUMBER`
+
+==== Type tests
+Additionally it may be practical to test the type of a value at runtime.
+
+*Example*
+`RETURN n.prop IS [NOT] NUMBER`
+
+==== Union type support
+Cypher's current type checker tracks sets of possible types for expressions.
+This could be understood as a form of union typing.
+Union types do not map easily to the type systems of most common target languages (Java, Javascript, ...), they complicate type checking, and are not required for type mapping to common external languages.
+This is why they have not been included in this proposal.
+They could be added instead at a later stage.
+
+*Example*
+`RETURN [1, "Yo"] :: LIST OF STRING | LIST OF NUMBER`
```

Comment:
Note: we have discussed a type checker that is able to track multiple specific types used within a list. It would probably still give lists a general contained type, but then tag specific offsets as having a more specific typed value. However nothing like that has been implemented.